### ISSUE TITLE
SC2-1-2-no-keyboard-trap-standard-navigation (Updated)

### ISSUE THREAD

**Issue by annethyme** [auto-wcag/auto-wcag#67]
---
name: No keyboard trap standard navigation
group:
- SC2-1-2-no-keyboard-trap-standard-navigation
- SC2-1-2-no-keyboard-trap-non-standard-navigation

description: |
  This rule checks if it is possible to use standard keyboard navigation to navigate through all content on a web page without becoming trapped in any element.

success_criterion: 
- 2.1.2 # No Keyboard Trap

test aspects:
- DOM Tree

authors:
- Dagfinn Rømen
- Geir Sindre Fossøy
- Malin Øvrebø
- Shadi Abou-Zahra
- Carlos Duarte
- Anne Thyme Nørregaard
- Stein Erik Skotkjerra
---

### Applicability

The rule applies to any HTML or SVG element on a web page that is [focusable][] by keyboard.

### Expectation

For each target element focus can cycle to the browser UI by using [standard keyboard navigation][].

**Note**: Cycling back to the browser UI can be done both by moving forward through the tab order and by moving backwards. It is not possible to fulfil this expectation by using browser specific shortcuts to return to the browser UI.

## Assumptions

- The WCAG success criterion applies to all content where focus can be moved to through keyboard navigation.
- It is possible to use unmodified arrow or tab keys, or other standard exit methods to move focus away.
- The focus order in keyboard navigation is cyclical, not linear, meaning that the focus order will cycle to the first/last element when it moves away from the last/first element.

## Accessibility support

There are no major accessibility support issues known for this rule.

## Background

- https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html
- https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/G21
- https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/F10

### Passed

```html
<a href ="#">Link 1</a>
<button class="target">Button1</button>
```

```html
<a href ="#" style="visibility: hidden;">Link 1</a>
<button class="target" style="visibility: hidden;">Button1</button>
```

```html
<div tabindex=“1”>Text</div>
<!-- something with tabindex=0 ? -->
```

### Failed

```html
<a href="#">Link 1</a>
<button class="target" onblur="setTimeout(() => this.focus(), 10)">Button1</button>
```

```html
<button class="target" onblur="setTimeout(() => this.nextSibling.focus(), 10)">Button1</button>
<button class="target" onblur="setTimeout(() => this.previousSibling.focus(), 10)">Button2</button>
```
### Inapplicable

```html
<h1>Page 1</h1>
```

```html
<a href="#" tabindex="-1">Tab key cannot reach here!</a>
<!-- Element you can't reach -->
```
```html
<button type="button" disabled>Click Me!</button>
<!-- Disabled element -->
```

```html
<button type="button" style=“display: none;”>Click Me!</button>
<!-- Hidden element -->
```

[focusable]: ../pages/algorithms/focusable.html
[standard keyboard navigation]: ../pages/algorithms/standard-keyboard-navigation.html

**Comment by WilcoFiers** [comment]
I suggest defining "focus into", or rephrasing it. What I'm guessing you mean is that it contains an element that's in the focus order?

About "focus out of". I think this can be misunderstood. If I open a lightbox, and that disabled all focusable elements in the background, and there is no way with keyboard to dismiss it, is this an element I can't "focus out of"?

It seems to me that the point of this SC is to make sure that, regardless of what state a page is in, you should never be prevented from accessing the browser UI.

**Comment by nitedog** [comment]
See also [related discussion on R2](https://github.com/auto-wcag/auto-wcag/issues/68#issuecomment-364650467).

The SC says "focus can be moved to" and "focus can be moved away from". Following that, here is a suggested rewrite:

Applicability: All components on a web page where focus can be moved to through keyboard navigation.
Expectation: Focus can be moved away from the component by using arrow keys, ESC key, TAB key, Shift+TAB or Enter key.

**Comment by WilcoFiers** [comment]
@nitedog Again, still not keen on the term "component". HTML doesn't have the concept of components, so you'll have to define it, otherwise it doesn't meet the requirement of being *objective*.

**Comment by nitedog** [comment]
WCAG 2.1 provides a [refined definition for "User Interface Components"](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components), which seems to include the term "Components" (see Note 3). It seems that "Component" is a short-hand for "User Interface Component". Does that help?

**Comment by WilcoFiers** [comment]
It helps in that "is perceived by the user as" clearly isn't objective. :)

**Comment by nitedog** [comment]
Is there a term you already use in other rules? This issue seems broader than this rule alone. I don't think we should define a counter term to what WCAG already uses. Instead, it may be better to seek clarification for this existing term (eg. in the Understanding documentation).

**Comment by WilcoFiers** [comment]
Most WCAG 2.0 definitions can't be used in rules. This is by design. WCAG 2.0 definitions are open-ended, so that they can be applied to future technologies. Where as definitions in rules should **never** be open-ended, because that's what causes all those interpretation problems in success criteria.

If you recall, one of the biggest concerns with moving to the applicability/expectation approach, was that it would be too open-ended. We don't want to just take the SC text and pour it into the rules format template. That doesn't get rid of any interpretation differences. To push rule writers to being more precise about this we've put in the latest ACT Rules format draft that applicability **must** be objective.

Ask yourself: What about this definition can I express objectively? The advantage of this is that you're not technology agnostic. Most of the rules do things like give a list of node names, or roles, or certain attributes to look for. Essentially you create a whitelist of everything you know to fit within that definition. That's not easy, and sometimes that's a little painful too because you'll have to ignore things you know to be issues, because you can't define them well enough.

**Comment by nitedog** [comment]
So again I ask, is there already a term being used in other rules? It seems that this term -- whatever it is called and however it is defined -- will be needed in several rules. Is there a common glossary of terms for ACT Rules?

**Comment by WilcoFiers** [comment]
This is the only thing that comes close:
https://github.com/auto-wcag/auto-wcag/blob/master/pages/algorithms/interactive-element.md

I don't think that would work for this rule though, since you're not trying to prevent focus from being trapped on a single element, but within a group of elements.

**Comment by nitedog** [comment]
I think it applies to single and group of objects/elements/components.

Actually, what is the definition of element, is it the the [HTML definition](https://www.w3.org/TR/html51/fullindex.html#index-elements)?

**Comment by WilcoFiers** [comment]
Yeah, HTML / ARIA / DOM / CSS terms would work.

**Comment by nitedog** [comment]
I updated the [proposal at top of this thread](https://github.com/auto-wcag/auto-wcag/issues/67#issue-295568469) using the term "element" instead of "component".

**Comment by geirsf** [comment]
We have updated the format based on the format here : #66

**Comment by DagfinnRomen** [comment]
I can understand the reasoning behind using the term "element". I still think that "component" could be more suited, because it also includes focusable content that is not written in HTML.

**Comment by WilcoFiers** [comment]
Ok, so this rule would fail things like the following:

```html
<button onblur="setTimeout(() => this.focus(), 10)">Evil</button>
```

But it doesn't actually fail the following:
```html
<button onblur="setTimeout(() => this.nextSibling.focus(), 10)">Evil1 </button>
<button onblur="setTimeout(() => this.previousSibling.focus(), 10)">Evil2 </button>
```

Any thoughts on how we could do that?

**Comment by MaliinO** [comment]
We have updated the Expectation part in the proposal to try to cover the situation commented by Wilco. Any thoughts?

**Comment by geirsf** [comment]
Wilco's comment: I don't think the logic of this works. Firstly because the "next element on the page" can be any element, I was going to suggest the "next element in the focus order", but I don't think that's true either. In things like ARIA tabs the next tab shouldn't be in the focus order. You should make sure this rule doesn't become a test that checks that the focus order makes sense. That's entirely different SC, and not the problem you are trying to describe.

This one is difficult to review without knowing what is going to be in the R2 rule.

**Comment by jribbens** [comment]
This code seems relevant: https://github.com/jquery/jquery-ui/blob/master/ui/focusable.js

**Comment by EmmaJP** [comment]
Perhaps the expectation shouldn't define what focus is moving to, only that it can be moved away from where is currently is. Something like...

For each target HTML element, or subsection of HTML elements, focus can be moved away from that element/s by using one or more of the following:

**Comment by annethyme** [comment]
Dagfinn, Geir Sindre, Stein Erik and Anne changed the whole Applicability + Expectations.

Old version that we have replaced: 

### Applicability

The rule applies to any HTML elements and subsections of HTML elements on a web page where focus can be moved to through keyboard navigation.

### Expectation

For each target HTML element, or subsection of HTML elements, focus can be moved away to an HTML element that has not previously had focus, by using one or more of the following:
- Tab key
- Shift+Tab
- Arrow keys
- Esc key
- Enter key

Please provide feedback on this new approach (see first comment in thread).

**Comment by annethyme** [comment]
Checked in with Kasper Isager, and we suggest editing the Applicability a bit: 

"The rule applies to any HTML or SVG element on a web page that is [focusable][] by keyboard." -->
"The rule applies to any HTML or SVG element on a web page that is [focusable][] and reachable through [sequential focus navigation] (https://www.w3.org/TR/html/editing.html#sequential-focus-navigation)."

"by keyboard" is suggested removed because we found this normative definition of "focusable" that says that focusable means that the element "can become the target of keyboard input".

There's also this spec for what it means that something can be focused: https://www.w3.org/TR/html/editing.html#can-be-focused

**Comment by annethyme** [comment]
@geirsf, Did you see the suggestion above about editing the Applicability? 

I'm also wondering if we should add more test cases. Don't know how many we need, but I've started thinking of some: 

### Passed

```html
<div tabindex=“1”>Text</a>
```

```html
<!-- something with tabindex=0 ? -->
```

### Inapplicable
```html
<a href="#" tabindex="-1">Tab key cannot reach here!</a>
```

```html
<button type="button" disabled>Click Me!</button>
```

```html
<button type="button" style=“display: none;”>Click Me!</button>
```

```html
<!-- something with tabindex=0 ? -->
```

**Comment by geirsf** [comment]
Added the new applicability from Anna to this and #68

**Comment by annethyme** [comment]
Made into pull request: https://github.com/auto-wcag/auto-wcag/pull/142

**Comment by annethyme** [comment]
Made into pull request: https://github.com/auto-wcag/auto-wcag/pull/142
